lib/common/asar.js

```diff
@@ -1,20 +1,311 @@
+            !integrityConfig.checksums ||
+            Object.keys(integrityConfig.checksums).length < 2
+          ) {
+            throwErrorAndExit(
+              `Invalid integrity config. No or too few checksums.\nraw integrity config: ${rawIntegrityString}`,
+              'EDAMAGEDAPP'
+            )
+            return
+          }
+
+          if (LOCAL_DEBUG) {
+            console.log(`integrity we're using:\n${rawIntegrityString}`)
+          }
+        }
+
+        // NOTE externals are always allowed, because we mostly have a folder inside our
+        // HOME_DIR for configuration and further packages (see XNotar).
+        /* let resourcesPath = process.resourcesPath;
+        if (!filePath.startsWith(resourcesPath)) {
+          if (integrityConfig.externalAllowed === true) {
+            return;
+          }
+          throwErrorAndExit(
+            `${filePath} not located in the application resources path (${resourcesPath}) but loading of external files is disallowed (externalAllowed: ${
+              integrityConfig.externalAllowed
+            })`,
+            "EDAMAGEDAPP"
+          );
+        } */
+
+        // find expected checksum by filename
+        const basename = path.basename(filePath)
+        const expected = integrityConfig.checksums[basename]
+
+        if (!expected) {
+          // NOTE in the renderer-process the "asar.getIntegrity" returns an empty list of checksums, because the bundle cannot be loaded.
+          // always fail for asar files (that's the minimum to be specified)
+          if (filePath.substr(-5) === '.asar') {
+            throwErrorAndExit(
+              `Application is corrupted, checksum not specified for ${filePath}\nIntegrities: ${JSON.stringify(
+                integrityConfig
+              )}`,
+              'EDAMAGEDAPP'
+            )
+          }
+
+          if (LOCAL_DEBUG) {
+            console.log(
+              `no checksum for external file ${basename}. That's okay.`
+            )
+          }
+
+          // return with no further action, if its a normal file with no checksum specified.
+          return
+        }
+
+        if (checked[basename] === true) {
+          // we've already checked this file.
+          if (LOCAL_DEBUG) {
+            console.log(`Already checked ${basename}`)
+          }
+          return
+        }
+
+        // is file already open?
+        const fd = fileFd ? fileFd.getFd() : originalFs.openSync(filePath, 'r')
+        if (fd < 0) {
+          const error = new Error(`ENOENT, ${filePath} not found`)
+          error.code = 'ENOENT'
+          error.errno = -2
+          throw error
+        }
+
+        const actual = hashSync(fd)
+
+        // close file, if no FD was given
+        if (!fileFd) {
+          originalFs.closeSync(fd)
+        }
+
+        if (expected !== actual) {
+          // set checked to true for this file to avoid permanent checks with failures here!
+          checked[basename] = true
```

Comment:
this seems fragile - allows subsequent reads to succeed

Reply:
Should be enough security IMHO. If you restart the application all checks are made again. Changes are only passing, if made at runtime after a check succeeded. Additionally, the "app.asar" can get a huge size, which will be checked over and over again, if this line is removed.

You can enable the logs locally, if you wanna see the requests to app.asar: https://github.com/procilon/electron/blob/8b097e824d8ce82f686aebc956d5d69798325189/lib/common/asar.js#L306